src/main/java/com/google/sps/data/EventStorage.java

```diff
@@ -86,15 +95,35 @@ public Event getEvent(String eventId) {
         continue;
       }
 
-      String category = (String) entity.getProperty("category");
+      List<String> tags = (List<String>) entity.getProperty("tags");
 
-      if (!search.eventInCategory(category)) {
+      int numberOfMatchingTags = search.countMatchingTags(tags);
+
+      if (numberOfMatchingTags == 0) {
         continue;
       }
 
-      events.add(Event.fromDatastoreEntity(entity));
+      events.put(Event.fromDatastoreEntity(entity), numberOfMatchingTags);
     }
     
+    return orderEventsByTags(events);
+  }
+
+  public List<Event> orderEventsByTags(Map<Event, Integer> eventsWithCounts) {
+    List<Event> events = new ArrayList<>();
+    List<Integer> counts = new ArrayList<>(eventsWithCounts.values()).stream() 
+                                      .distinct().collect(Collectors.toList());
+
+    counts.sort(Comparator.reverseOrder());
+
+    for (Integer count : counts) {
+      for (Map.Entry<Event, Integer> eventWithCount : eventsWithCounts.entrySet()) {
+        if (eventWithCount.getValue().equals(count)) {
+          events.add(eventWithCount.getKey());
+        }
+      }
+    }
+
     return events;
   }
```

Comment:
This looks a bit hacky. Maybe have a look here how to sort a map by value: https://dzone.com/articles/how-to-sort-a-map-by-value-in-java-8